### ISSUE TITLE
import_piefxis fails for vcards with photo elements

### ISSUE THREAD

**Issue by michaelkuhn** [processone/ejabberd#3166]
## Environment

- ejabberd version: 19.09.1
- Erlang version: Erlang (SMP,ASYNC_THREADS,HIPE) (BEAM) emulator version 10.6.2
- OS: Linux (Fedora)
- Installed from: distro package

## Errors from error.log/crash.log

`[error] <0.727.0>@ejabberd_piefxis:stop:554 Unexpected tag: {xmlel, ...`

## Bug description

I tried importing a piefxis XML (exported by ejabberd) with the following vcard: `<vCard xmlns='vcard-temp'><PHOTO><BINVAL>...</BINVAL><TYPE>image/png</TYPE></PHOTO><NICKNAME>...</NICKNAME></vCard>`, which failed with the above error. Removing the PHOTO element fixed the problem.

**Comment by badlop** [comment]
Hi, I can't reproduce that problem. I installed ejabberd 19.09.1 from source code, then created an account and set some avatar, then exported piefxis and this is the result:

file 20200210-164255.xml
`<?xml version='1.0' encoding='UTF-8'?>
<server-data xmlns='urn:xmpp:pie:0' xmlns:xi='http://www.w3.org/2001/XInclude'>
<xi:include href='20200210-164255_localhost.xml'/></server-data>`

file 20200210-164255_localhost.xml with the BINVAL cropped to just a few characters:
```
<?xml version='1.0' encoding='UTF-8'?>
<host xmlns='urn:xmpp:pie:0' xmlns:xi='http://www.w3.org/2001/XInclude' jid='localhost'>
<user name='user1' password='mypass11'>
<vCard xmlns='vcard-temp'><PHOTO>
<BINVAL>R0lGOw==</BINVAL>
<TYPE>image/png</TYPE></PHOTO><NICKNAME>mi apodo</NICKNAME></vCard></user></host>
```

Then I stop ejabberd, delete the database, start it again, and try to import:
`ejabberdctl import_piefxis /tmp/asd/20200210-164255.xml `

It shows no error in the log files, and in fact I see the vcard element in the mnesia database.

Can you show a minimal file that throws the error when imported, so that I can debug it? Or try my example, does it work correctly, or also crashes in your installation?

**Comment by michaelkuhn** [comment]
Huh, strange. Could this be a size problem. As far as I remember, the photo was a bit larger (~20 KB within the XML). I'll try to dig out the actual XML that produced the error.

**Comment by michaelkuhn** [comment]
I was able to reproduce the problem by extending your example a bit: https://ikkoku.de/ejabberd.tar.gz

Using these XML files I get `ejabberd_piefxis:stop:554 Unexpected tag: {xmlel,<<"user">>,[{<<"name">>,<<"user1">>},{<<"password">>,<<"mypass11">>}],[{xmlcdata,<<"\n">>},{xmlel,<<"vCard">>,[{<<"xmlns">>,<<"vcard-temp">>}],[{xmlel,<<"PHOTO">>,[],[{xmlcdata,<<"\n">>},{xmlel,<<"BINVAL">>,[],[{xmlcdata,<<"RR0lGOw==` etc.

If I remove the second vhost, I get `@ejabberd_piefxis:stop:554 failed to decode XML '<vCard xmlns='vcard-temp'><P
HOTO>
<BINVAL>RR0lGOw==...': Bad value of cdata in tag <BINVAL/> qualified by namespace 'vcard-temp'`

Not really sure what is going on there.

**Comment by badlop** [comment]
> ejabberd_piefxis:stop:554 Unexpected tag: {xmlel,<<"user">>,

Aha, now I see the problem too, and it's easy to reproduce. This happens when importing several vhosts. Only one of them is correctly imported, with its users. When importing users of the second vhost, it crashes as you reported.

If you have few vhosts, a workaround is to edit 20200210-164255.xml to mention only on vhost file, import it, and then repeat for each vhost.

> If I remove the second vhost, I get @ejabberd_piefxis:stop:554 failed to decode XML '<vCard xmlns='vcard-temp'><P HOTO> <BINVAL>RR0lGOw==...': Bad value of cdata in tag <

Well, because the CDATA in BINVAL is invalid :)

**Comment by mremond** [comment]
@badlop How hard is it to improve the importer to support multiple vhosts?

**Comment by michaelkuhn** [comment]
> Aha, now I see the problem too, and it's easy to reproduce. This happens when importing several vhosts. Only one of them is correctly imported, with its users. When importing users of the second vhost, it crashes as you reported.

Actually, the problem happened with the **first** vhost for me. However, if I remove the second one, the first one imports correctly. Also, after removing the photo element, both imported correctly, which is why I assumed the problem to be the photo element.

**Comment by prefiks** [comment]
This issue should be fixed by commit a19fdb717bc320a41092195a97bbd2281ae2472e

**Comment by prefiks** [comment]
This will only fix problem with parsing multiple host files, this will not fix importing photo that was stored in that tgz file, it's not valid (that's not properly base64 encoded data) and ejabberd will reject it because of this.
